`Queue[Id(1,2,d3d12)] does not exist` internal panic during `Queue::write_buffer()` on DX12 backend

**Description**
I observed the following panic as an intermittent test failure in my own project's CI when attempting to update to wgpu 0.19. The tests share one `Adapter` between test threads, but each test uses its own `Device` and `Queue`. I don't know if the failure is Windows-specific as it has only happened once so far.

```rust
thread 'tests::scale_to_integer_step_test' panicked at ...\wgpu-core-0.19.0\src\storage.rs:113:39:
Queue[Id(1,2,d3d12)] does not exist
stack backtrace:
...
    2: wgpu_core::storage::Storage<wgpu_core::pipeline::ShaderModule<wgpu_hal::gles::Api>,wgpu_core::id::Id<wgpu_core::pipeline::ShaderModule<wgpu_hal::empty::Api> > >::get<wgpu_core::pipeline::ShaderModule<wgpu_hal::gles::Api>,wgpu_core::id::Id<wgpu_core::pipel
              at ...\wgpu-core-0.19.0\src\storage.rs:113
    3: wgpu_core::registry::Registry<wgpu_core::id::Id<wgpu_core::pipeline::ComputePipeline<wgpu_hal::empty::Api> >,wgpu_core::pipeline::ComputePipeline<wgpu_hal::dx12::Api> >::get<wgpu_core::id::Id<wgpu_core::pipeline::ComputePipeline<wgpu_hal::empty::Api> >,wg
              at ...\wgpu-core-0.19.0\src\registry.rs:136
    4: wgpu_core::global::Global<wgpu_core::identity::IdentityManagerFactory>::queue_write_buffer<wgpu_core::identity::IdentityManagerFactory,wgpu_hal::dx12::Api>
              at ...\wgpu-core-0.19.0\src\device\queue.rs:377
    5: wgpu::backend::wgpu_core::impl$7::queue_write_buffer
              at ...\wgpu-0.19.0\src\backend\wgpu_core.rs:2092
    6: wgpu::context::impl$5::queue_write_buffer<wgpu::backend::wgpu_core::ContextWgpuCore>
              at ...\wgpu-0.19.0\src\context.rs:2911
    7: wgpu::Queue::write_buffer
              at ...\wgpu-0.19.0\src\lib.rs:4582
...
```

**Repro**
No standalone repro yet.

**Platform**
wgpu: 0.19.0

GitHub Actions WIndows runner:
```
Current runner version: '2.311.0'
Operating System
  Microsoft Windows Server [2](https://github.com/kpreid/all-is-cubes/actions/runs/7565119937/job/20600346191?pr=449#step:1:2)022
  10.0.20[3](https://github.com/kpreid/all-is-cubes/actions/runs/7565119937/job/20600346191?pr=449#step:1:3)[4](https://github.com/kpreid/all-is-cubes/actions/runs/7565119937/job/20600346191?pr=449#step:1:4)8
  Datacenter
Runner Image
  Image: windows-2022
  Version: 2024010[8](https://github.com/kpreid/all-is-cubes/actions/runs/7565119937/job/20600346191?pr=449#step:1:9).1.0
  Included Software: https://github.com/actions/runner-images/blob/win22/20240108.1/images/windows/Windows2022-Readme.md
  Image Release: https://github.com/actions/runner-images/releases/tag/win22%2F20240[10](https://github.com/kpreid/all-is-cubes/actions/runs/7565119937/job/20600346191?pr=449#step:1:11)8.1
```

I have something like that:
```
21:11:24 [ERROR] Uncaptured device error: Validation { source: ContextError { string: "Queue::write_buffer", cause: Queue(InvalidQueueId), label_key: "", label: "" }, description: "Validation Error\n\nCaused by:\n    In Queue::write_buffer\n    QueueId is invalid\n" }
21:11:24 [ERROR] Uncaptured device error: Validation { source: ContextError { string: "Queue::write_buffer", cause: Queue(InvalidQueueId), label_key: "", label: "" }, description: "Validation Error\n\nCaused by:\n    In Queue::write_buffer\n    QueueId is invalid\n" }
21:11:24 [ERROR] Uncaptured device error: Validation { source: ContextError { string: "Queue::write_buffer", cause: Queue(InvalidQueueId), label_key: "", label: "" }, description: "Validation Error\n\nCaused by:\n    In Queue::write_buffer\n    QueueId is invalid\n" }
21:11:25 [ERROR] thread '<unnamed>' panicked at 'Error in Queue::submit: Validation Error

Caused by:
    QueueId is invalid
': C:\Users\Eddy\.cargo\git\checkouts\wgpu-53e70f8674b08dd4\adf1e3b\wgpu\src\backend\wgpu_core.rs:2228
   0: backtrace::backtrace::trace
   1: backtrace::capture::Backtrace::new
   2: <backtrace::capture::Backtrace as core::default::Default>::default
   3: <log_panics::Shim as core::fmt::Debug>::fmt
   4: alloc::boxed::impl$49::call
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library\alloc\src\boxed.rs:2021
   5: std::panicking::rust_panic_with_hook
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library\std\src\panicking.rs:783
   6: std::panicking::begin_panic_handler::closure$0
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library\std\src\panicking.rs:657
   7: std::sys_common::backtrace::__rust_end_short_backtrace<std::panicking::begin_panic_handler::closure_env$0,never$>
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library\std\src\sys_common\backtrace.rs:170
   8: std::panicking::begin_panic_handler
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library\std\src\panicking.rs:645
   9: core::panicking::panic_fmt
             at /rustc/82e1608dfa6e0b5569232559e3d385fea5a93112/library\core\src\panicking.rs:72
  10: <wgpu::backend::wgpu_core::ContextWgpuCore as core::fmt::Debug>::fmt
  11: <T as wgpu::context::DynContext>::queue_submit
  ```
  
[trace.zip](https://github.com/gfx-rs/wgpu/files/13981666/trace.zip)

hope this helps

--- a/player/src/bin/play.rs
+++ b/player/src/bin/play.rs
@@ -61,7 +61,7 @@ fn main() {
     }
     .unwrap();
 
-    let device = match actions.pop() {
+    let (device, queue) = match actions.pop() {
         Some(trace::Action::Init { desc, backend }) => {
             log::info!("Initializing the device for backend: {:?}", backend);
             let adapter = global
@@ -80,18 +80,19 @@ fn main() {
 
             let info = gfx_select!(adapter => global.adapter_get_info(adapter)).unwrap();
             log::info!("Picked '{}'", info.name);
-            let id = wgc::id::Id::zip(1, 0, backend);
+            let device_id = wgc::id::Id::zip(1, 0, backend);
+            let queue_id = wgc::id::Id::zip(1, 0, backend);
             let (_, _, error) = gfx_select!(adapter => global.adapter_request_device(
                 adapter,
                 &desc,
                 None,
-                Some(id),
-                Some(id.into_queue_id())
+                Some(device_id),
+                Some(queue_id)
             ));
             if let Some(e) = error {
                 panic!("{:?}", e);
             }
-            id
+            (device_id, queue_id)
         }
         _ => panic!("Expected Action::Init"),
     };
@@ -102,7 +103,7 @@ fn main() {
         gfx_select!(device => global.device_start_capture(device));
 
         while let Some(action) = actions.pop() {
-            gfx_select!(device => global.process(device, action, &dir, &mut command_buffer_id_manager));
+            gfx_select!(device => global.process(device, queue, action, &dir, &mut command_buffer_id_manager));
         }
 
         gfx_select!(device => global.device_stop_capture(device));
@@ -156,7 +157,7 @@ fn main() {
                                 target.exit();
                         }
                         Some(action) => {
-                            gfx_select!(device => global.process(device, action, &dir, &mut command_buffer_id_manager));
+                            gfx_select!(device => global.process(device, queue, action, &dir, &mut command_buffer_id_manager));
                         }
                         None => {
                             if !done {

--- a/player/src/lib.rs
+++ b/player/src/lib.rs
@@ -16,6 +16,7 @@ pub trait GlobalPlay {
     fn process<A: wgc::hal_api::HalApi>(
         &self,
         device: wgc::id::DeviceId,
+        queue: wgc::id::QueueId,
         action: trace::Action,
         dir: &Path,
         comb_manager: &mut wgc::identity::IdentityManager<wgc::id::markers::CommandBuffer>,
@@ -131,6 +132,7 @@ impl GlobalPlay for wgc::global::Global {
     fn process<A: wgc::hal_api::HalApi>(
         &self,
         device: wgc::id::DeviceId,
+        queue: wgc::id::QueueId,
         action: trace::Action,
         dir: &Path,
         comb_manager: &mut wgc::identity::IdentityManager<wgc::id::markers::CommandBuffer>,
@@ -327,7 +329,7 @@ impl GlobalPlay for wgc::global::Global {
                 let bin = std::fs::read(dir.join(data)).unwrap();
                 let size = (range.end - range.start) as usize;
                 if queued {
-                    self.queue_write_buffer::<A>(device.into_queue_id(), id, range.start, &bin)
+                    self.queue_write_buffer::<A>(queue, id, range.start, &bin)
                         .unwrap();
                 } else {
                     self.device_set_buffer_data::<A>(id, range.start, &bin[..size])
@@ -341,11 +343,11 @@ impl GlobalPlay for wgc::global::Global {
                 size,
             } => {
                 let bin = std::fs::read(dir.join(data)).unwrap();
-                self.queue_write_texture::<A>(device.into_queue_id(), &to, &bin, &layout, &size)
+                self.queue_write_texture::<A>(queue, &to, &bin, &layout, &size)
                     .unwrap();
             }
             Action::Submit(_index, ref commands) if commands.is_empty() => {
-                self.queue_submit::<A>(device.into_queue_id(), &[]).unwrap();
+                self.queue_submit::<A>(queue, &[]).unwrap();
             }
             Action::Submit(_index, commands) => {
                 let (encoder, error) = self.device_create_command_encoder::<A>(
@@ -361,8 +363,7 @@ impl GlobalPlay for wgc::global::Global {
                     panic!("{e}");
                 }
                 let cmdbuf = self.encode_commands::<A>(encoder, commands);
-                self.queue_submit::<A>(device.into_queue_id(), &[cmdbuf])
-                    .unwrap();
+                self.queue_submit::<A>(queue, &[cmdbuf]).unwrap();
             }
         }
     }

--- a/player/tests/test.rs
+++ b/player/tests/test.rs
@@ -106,6 +106,7 @@ impl Test<'_> {
     ) {
         let backend = adapter.backend();
         let device_id = wgc::id::Id::zip(test_num, 0, backend);
+        let queue_id = wgc::id::Id::zip(test_num, 0, backend);
         let (_, _, error) = wgc::gfx_select!(adapter => global.adapter_request_device(
             adapter,
             &wgt::DeviceDescriptor {
@@ -116,7 +117,7 @@ impl Test<'_> {
             },
             None,
             Some(device_id),
-            Some(device_id.into_queue_id())
+            Some(queue_id)
         ));
         if let Some(e) = error {
             panic!("{:?}", e);
@@ -125,7 +126,7 @@ impl Test<'_> {
         let mut command_buffer_id_manager = wgc::identity::IdentityManager::new();
         println!("\t\t\tRunning...");
         for action in self.actions {
-            wgc::gfx_select!(device_id => global.process(device_id, action, dir, &mut command_buffer_id_manager));
+            wgc::gfx_select!(device_id => global.process(device_id, queue_id, action, dir, &mut command_buffer_id_manager));
         }
         println!("\t\t\tMapping...");
         for expect in &self.expectations {

--- a/tests/tests/device.rs
+++ b/tests/tests/device.rs
@@ -1,6 +1,8 @@
 use std::sync::atomic::AtomicBool;
 
-use wgpu_test::{fail, gpu_test, FailureCase, GpuTestConfiguration, TestParameters};
+use wgpu_test::{
+    fail, gpu_test, FailureCase, GpuTestConfiguration, TestParameters, TestingContext,
+};
 
 #[gpu_test]
 static CROSS_DEVICE_BIND_GROUP_USAGE: GpuTestConfiguration = GpuTestConfiguration::new()
@@ -907,4 +909,27 @@ static DEVICE_DESTROY_THEN_BUFFER_CLEANUP: GpuTestConfiguration = GpuTestConfigu
 
         // Poll the device, which should try to clean up its resources.
         ctx.instance.poll_all(true);
+    });
+
+#[gpu_test]
+static DEVICE_AND_QUEUE_HAVE_DIFFERENT_IDS: GpuTestConfiguration = GpuTestConfiguration::new()
+    .parameters(TestParameters::default())
+    .run_async(|ctx| async move {
+        let TestingContext {
+            adapter,
+            device_features,
+            device_limits,
+            device,
+            queue,
+            ..
+        } = ctx;
+
+        drop(device);
+
+        let (device2, queue2) =
+            wgpu_test::initialize_device(&adapter, device_features, device_limits).await;
+
+        drop(queue);
+        drop(device2);
+        drop(queue2); // this would previously panic since we would try to use the Device ID to drop the Queue
     });

--- a/wgpu-core/src/device/global.rs
+++ b/wgpu-core/src/device/global.rs
@@ -7,7 +7,7 @@ use crate::{
         ResolvedBindGroupEntry, ResolvedBindingResource, ResolvedBufferBinding,
     },
     command, conv,
-    device::{bgl, life::WaitIdleError, queue, DeviceError, DeviceLostClosure, DeviceLostReason},
+    device::{bgl, life::WaitIdleError, DeviceError, DeviceLostClosure, DeviceLostReason},
     global::Global,
     hal_api::HalApi,
     id::{self, AdapterId, DeviceId, QueueId, SurfaceId},
@@ -2040,7 +2040,7 @@ impl Global {
     pub fn device_poll<A: HalApi>(
         &self,
         device_id: DeviceId,
-        maintain: wgt::Maintain<queue::WrappedSubmissionIndex>,
+        maintain: wgt::Maintain<crate::SubmissionIndex>,
     ) -> Result<bool, WaitIdleError> {
         api_log!("Device::poll {maintain:?}");
 
@@ -2049,15 +2049,6 @@ impl Global {
             .devices
             .get(device_id)
             .map_err(|_| DeviceError::InvalidDeviceId)?;
-
-        if let wgt::Maintain::WaitForSubmissionIndex(submission_index) = maintain {
-            if submission_index.queue_id != device_id.into_queue_id() {
-                return Err(WaitIdleError::WrongSubmissionIndex(
-                    submission_index.queue_id,
-                    device_id,
-                ));
-            }
-        }
 
         let DevicePoll {
             closures,
@@ -2071,7 +2062,7 @@ impl Global {
 
     fn poll_single_device<A: HalApi>(
         device: &crate::device::Device<A>,
-        maintain: wgt::Maintain<queue::WrappedSubmissionIndex>,
+        maintain: wgt::Maintain<crate::SubmissionIndex>,
     ) -> Result<DevicePoll, WaitIdleError> {
         let snatch_guard = device.snatchable_lock.read();
         let fence = device.fence.read();

--- a/wgpu-core/src/device/life.rs
+++ b/wgpu-core/src/device/life.rs
@@ -4,7 +4,6 @@ use crate::{
         DeviceError, DeviceLostClosure,
     },
     hal_api::HalApi,
-    id,
     resource::{self, Buffer, Texture, Trackable},
     snatch::SnatchGuard,
     SubmissionIndex,
@@ -112,8 +111,8 @@ impl<A: HalApi> ActiveSubmission<A> {
 pub enum WaitIdleError {
     #[error(transparent)]
     Device(#[from] DeviceError),
-    #[error("Tried to wait using a submission index from the wrong device. Submission index is from device {0:?}. Called poll on device {1:?}.")]
-    WrongSubmissionIndex(id::QueueId, id::DeviceId),
+    #[error("Tried to wait using a submission index ({0}) that has not been returned by a successful submission (last successful submission: {1})")]
+    WrongSubmissionIndex(SubmissionIndex, SubmissionIndex),
     #[error("GPU got stuck :(")]
     StuckGpu,
 }

--- a/wgpu-core/src/device/queue.rs
+++ b/wgpu-core/src/device/queue.rs
@@ -131,13 +131,6 @@ impl SubmittedWorkDoneClosure {
     }
 }
 
-#[repr(C)]
-#[derive(Debug, Copy, Clone)]
-pub struct WrappedSubmissionIndex {
-    pub queue_id: QueueId,
-    pub index: SubmissionIndex,
-}
-
 /// A texture or buffer to be freed soon.
 ///
 /// This is just a tagged raw texture or buffer, generally about to be added to
@@ -1044,7 +1037,7 @@ impl Global {
         &self,
         queue_id: QueueId,
         command_buffer_ids: &[id::CommandBufferId],
-    ) -> Result<WrappedSubmissionIndex, QueueSubmitError> {
+    ) -> Result<SubmissionIndex, QueueSubmitError> {
         profiling::scope!("Queue::submit");
         api_log!("Queue::submit {queue_id:?}");
 
@@ -1351,10 +1344,7 @@ impl Global {
 
         api_log!("Queue::submit to {queue_id:?} returned submit index {submit_index}");
 
-        Ok(WrappedSubmissionIndex {
-            queue_id,
-            index: submit_index,
-        })
+        Ok(submit_index)
     }
 
     pub fn queue_get_timestamp_period<A: HalApi>(

--- a/wgpu-core/src/device/resource.rs
+++ b/wgpu-core/src/device/resource.rs
@@ -55,8 +55,8 @@ use std::{
 };
 
 use super::{
-    queue::{self, Queue},
-    DeviceDescriptor, DeviceError, UserClosures, ENTRYPOINT_FAILURE_ERROR, ZERO_BUFFER_SIZE,
+    queue::Queue, DeviceDescriptor, DeviceError, UserClosures, ENTRYPOINT_FAILURE_ERROR,
+    ZERO_BUFFER_SIZE,
 };
 
 /// Structure describing a logical device. Some members are internally mutable,
@@ -407,7 +407,7 @@ impl<A: HalApi> Device<A> {
     pub(crate) fn maintain<'this>(
         &'this self,
         fence_guard: crate::lock::RwLockReadGuard<Option<A::Fence>>,
-        maintain: wgt::Maintain<queue::WrappedSubmissionIndex>,
+        maintain: wgt::Maintain<crate::SubmissionIndex>,
         snatch_guard: SnatchGuard,
     ) -> Result<(UserClosures, bool), WaitIdleError> {
         profiling::scope!("Device::maintain");
@@ -417,9 +417,20 @@ impl<A: HalApi> Device<A> {
         // Determine which submission index `maintain` represents.
         let submission_index = match maintain {
             wgt::Maintain::WaitForSubmissionIndex(submission_index) => {
-                // We don't need to check to see if the queue id matches
-                // as we already checked this from inside the poll call.
-                submission_index.index
+                let last_successful_submission_index = self
+                    .last_successful_submission_index
+                    .load(Ordering::Acquire);
+
+                if let wgt::Maintain::WaitForSubmissionIndex(submission_index) = maintain {
+                    if submission_index > last_successful_submission_index {
+                        return Err(WaitIdleError::WrongSubmissionIndex(
+                            submission_index,
+                            last_successful_submission_index,
+                        ));
+                    }
+                }
+
+                submission_index
             }
             wgt::Maintain::Wait => self
                 .last_successful_submission_index

--- a/wgpu-core/src/id.rs
+++ b/wgpu-core/src/id.rs
@@ -326,12 +326,6 @@ impl CommandBufferId {
     }
 }
 
-impl DeviceId {
-    pub fn into_queue_id(self) -> QueueId {
-        Id(self.0, PhantomData)
-    }
-}
-
 #[test]
 fn test_id_backend() {
     for &b in &[

--- a/wgpu-core/src/lib.rs
+++ b/wgpu-core/src/lib.rs
@@ -102,7 +102,7 @@ pub(crate) use hash_utils::*;
 /// The index of a queue submission.
 ///
 /// These are the values stored in `Device::fence`.
-type SubmissionIndex = hal::FenceValue;
+pub type SubmissionIndex = hal::FenceValue;
 
 type Index = u32;
 type Epoch = u32;

--- a/wgpu/src/backend/wgpu_core.rs
+++ b/wgpu/src/backend/wgpu_core.rs
@@ -552,7 +552,7 @@ impl crate::Context for ContextWgpuCore {
     type SurfaceId = wgc::id::SurfaceId;
     type SurfaceData = Surface;
     type SurfaceOutputDetail = SurfaceOutputDetail;
-    type SubmissionIndexData = wgc::device::queue::WrappedSubmissionIndex;
+    type SubmissionIndexData = wgc::SubmissionIndex;
 
     type RequestAdapterFuture = Ready<Option<(Self::AdapterId, Self::AdapterData)>>;
 
@@ -666,7 +666,7 @@ impl crate::Context for ContextWgpuCore {
             id: queue_id,
             error_sink,
         };
-        ready(Ok((device_id, device, device_id.into_queue_id(), queue)))
+        ready(Ok((device_id, device, queue_id, queue)))
     }
 
     fn instance_poll_all_devices(&self, force_wait: bool) -> bool {

